### ISSUE TITLE
Clarification needed on dense acumulation

### ISSUE THREAD

**Issue by karam-nus** [facebookresearch/DeltaCNN#1]
In section 2.1 :
"The dense results are only accumulated at the final layer. This way, we also accelerate non-convolutional layers like pooling, up-sampling and activations."

This seems contrary to Figure 2 and point just next to it:

"Furthermore, we avoid switching between sparse and dense computations and only need to cache accumulated values for nonlinear layers, reducing the number of caches, without loss in accuracy."

In figure 2 we can see "Dense Accumulated Values" for "Sparse Activation", "Sparse Pooling" apart from "Output Accumulation".

Can you please elaborate? Is dense accumulation happening at non-linear layers as well? Thanks.

**Comment by dabeschte** [comment]
Hi!

Oh, I see. This sentence is a bit confusing.
What we intended to say is, that we only propagate sparse Delta values from layer to layer until the output layer, which then returns a dense tensor with accumulated values instead of Deltas.

For clarification:
- non-linear layers _store_ dense accumulated values, but the frame-by-frame update is sparse and only _processed_ at updated pixels.
- the output layer stores accumulated values the same way and only updated pixels are accumulated onto the buffer every frame. But it returns a dense copy of the new accumulated values buffer (which we also referred to as _previous output buffer_ for this layer).
- the reason why all non-linear layers need to store accumulated values is because they can't be processed on the sparse Delta directly. E.g. if a value in the feature map is -1 in the first frame and we apply ReLU activation, the result is 0. If in the next frame, we get a positive update of 0.5 for this value, ReLU would return 0.5, but actually, the accumulated value is still -0.5, therefore the correct result is 0 again - resulting in a Delta of (0-0)=0 on the output of the activation. This is why we store accumulated values, but we only have to look at updated pixels. the remaining pixels will result in a 0 output value.
